datanode/sqlstore/proposals.go

```diff
@@ -311,19 +386,40 @@ func (ps *Proposals) Get(ctx context.Context,
 	defer results.Close()
 
 	proposals := make([]entities.Proposal, 0)
+	fetchedBatches := map[entities.ProposalID]struct{}{}
 
 	for {
 		rows, err := results.Query()
 		if err != nil {
 			break
 		}
 
-		var props []entities.Proposal
-
-		if err := pgxscan.ScanAll(&props, rows); err != nil {
+		var matchedProps []entities.Proposal
+		if err := pgxscan.ScanAll(&matchedProps, rows); err != nil {
 			return nil, pageInfo, fmt.Errorf("querying proposals: %w", err)
 		}
 
+		var props []entities.Proposal
+		for _, p := range matchedProps {
+			var batchID entities.ProposalID
+			if p.BelongsToBatch() {
+				batchID = p.BatchID
+			} else if p.IsBatch() {
+				batchID = p.ID
+			}
+
+			if _, ok := fetchedBatches[batchID]; ok {
+				continue
+			}
+
+			p, err := ps.extendOrGetBatchProposal(ctx, p)
```

Comment:
The way this works seems very unintuitive to me. 

If I ask to list all proposals, we get given the batch proposals which contains a list of the batch terms, but the sub-proposals themselves are not returned. So I can't get to the ID of the sub-proposals.

If I _somehow_ did manage to work out the sub-proposal ID and called the `GetGovernanceData` with the sub-proposal ID, it then returns the batch-proposal (which consequently has a different ID) so it doesn't return what I asked.

The result is that I can't seem to find a way to get a sub-proposal from the governance data end-points. Which means I can't see the `state` field of any of them and thats important because it is possible for a proposal to fail between it passing and enacting.

I feel like @edd might have some opinions on what the API should look like.

Reply:
Ignore the above: I didn't expand the JSON enough and missed that for a batch proposal it also has a slice of all the sub-proposals, so you can see everything. I still think its a bit weird to ask for a proposal by ID and get a different proposal back, but its no the end of the world.